Add rust `edition` feature

MahadMuhammad opened the pull request:
- Fixes: https://github.com/Rust-GCC/rusttest-to-dg/issues/8



---
- Depends on: https://github.com/Rust-GCC/rusttest-to-dg/pull/10

MahadMuhammad committed:
Add `header` module

Signed-off-by: Muhammad Mahad <mahadtxt@gmail.com>

MahadMuhammad committed:
Add `HeaderLine` struct to store header info

Signed-off-by: Muhammad Mahad <mahadtxt@gmail.com>

MahadMuhammad committed:
Add `parse_edition` functions

Signed-off-by: Muhammad Mahad <mahadtxt@gmail.com>

MahadMuhammad committed:
Add function for `dejagnu` edition format

Signed-off-by: Muhammad Mahad <mahadtxt@gmail.com>

MahadMuhammad committed:
Add logic for tokenizing header line

MahadMuhammad committed:
Add function to store `HeaderLine` info

Signed-off-by: Muhammad Mahad <mahadtxt@gmail.com>

MahadMuhammad committed:
Add exhaustive names of all known directives

Signed-off-by: Muhammad Mahad <mahadtxt@gmail.com>

MahadMuhammad committed:
feat/Add rust edition functionality

Signed-off-by: Muhammad Mahad <mahadtxt@gmail.com>

CohenArthur commented on src/known-directives.rs:
```diff
@@ -0,0 +1,236 @@
+// Copied from https://github.com/rust-lang/rust/blob/master/src/tools/compiletest/src/command-list.rs
+pub const KNOWN_DIRECTIVE_NAMES: &[&str] = &[
+    // tidy-alphabetical-start
```
I think we can remove those comments as they are for an internal `rustc` tool I believe

CohenArthur commented on src/known-directives.rs:
```diff
@@ -0,0 +1,236 @@
+    "unset-rustc-env",
+    // Used by the tidy check `unknown_revision`.
+    "unused-revision-names",
+    // tidy-alphabetical-end
```
same here

MahadMuhammad committed:
remove rustc internal comments

Signed-off-by: Muhammad Mahad <mahadtxt@gmail.com>

MahadMuhammad committed:
Refactor header line condition

Signed-off-by: Muhammad Mahad <mahadtxt@gmail.com>

MahadMuhammad committed:
Refactor header line condition to avoid `None` unwrap

If the test source file doesn't contain any headers, or
that are not added till now then the header value
will be None and becomes panic

Signed-off-by: Muhammad Mahad <mahadtxt@gmail.com>

P-E-P commented on src/cli.rs:
```diff
@@ -65,6 +72,8 @@ mod tests {
     }
 
+    /// clap reports most development errors as `debug_assert!`s
+    /// See this for more details, [here](https://docs.rs/clap/4.5.15/clap/_derive/_tutorial/chapter_4/index.html)
```
I'm not sure a doc comment is required ? A regular comment might be enough

MahadMuhammad replied:
Converted to regular comments

you're right, these doc comments are not helpful for users, and should only kept for understanding internals

P-E-P commented on src/errors.rs:
```diff
@@ -1,10 +1,21 @@
+    std::{fmt, str::FromStr},
+};
+
+// https://docs.rs/once_cell/1.19.0/once_cell/#lazily-compiled-regex
```
```suggestion
/// Macro used to lazily create a new regex the first time it is invoked.
///
/// # Arguments
///
/// * `re` - The regex literal string used to build the automaton
///
/// # Example
///
/// ```rust
/// assert!(regex!(r"\w").is_match(" "));
/// ```
///
/// Taken from here https://docs.rs/once_cell/1.19.0/once_cell/#lazily-compiled-regex
```

Macros can be tricky to understand, I'd like this to be more thoroughly documented. A doctest example even provides a great insight (the example should compile & run when you launch cargo test).

MahadMuhammad replied:
Sure thing, thanks for the brief example :)

P-E-P commented on src/errors.rs:
```diff
@@ -60,11 +72,14 @@ pub struct Error {
     /// What kind of message we expect (e.g., warning, error, suggestion).
     /// `None` if not specified or unknown message kind.
     pub kind: Option<RustcErrorKind>,
+    ///Note: if we are loading this from rustc source file, this might be incomplete
```
```suggestion
    /// Note: if we are loading this from rustc source file, this might be incomplete
```

P-E-P commented on src/errors.rs:
```diff
@@ -60,11 +72,14 @@ pub struct Error {
 
 impl fmt::Display for Error {
+    /// Formats the `Error` for display according to `DejaGnu` format
+    /// See `DejaGnu` documentation [here](https://gcc.gnu.org/onlinedocs/gccint/testsuites/directives-used-within-dejagnu-tests/syntax-and-descriptions-of-test-directives.html)
```
```suggestion
    /// Formats the `Error` for display according to `DejaGnu` format
    /// See [`DejaGnu` documentation](https://gcc.gnu.org/onlinedocs/gccint/testsuites/directives-used-within-dejagnu-tests/syntax-and-descriptions-of-test-directives.html)
```

P-E-P commented on src/errors.rs:
```diff
@@ -60,11 +72,14 @@ pub struct Error {
+    /// See `DejaGnu` documentation [here](https://gcc.gnu.org/onlinedocs/gccint/testsuites/directives-used-within-dejagnu-tests/syntax-and-descriptions-of-test-directives.html)
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         use RustcErrorKind::*;
 
```
```rust
let error_type = match &self.kind {
            Some(Help) => "help",
            Some(Note) => "dg-note",
            Some(Suggestion) => "suggestion",
            Some(Warning) => "dg-warning",
            Some(Error) | None => "dg-error",
};
```

P-E-P commented on src/header.rs:
```diff
@@ -0,0 +1,96 @@
+        Some(HeaderLine {
+            line_number: line_number + 1, // 1 based-indexed instead of zero based
+            _directive: "edition",
+            dejagnu_header: to_dejagnu_edition(edition.unwrap().as_str()),
```
Why can we unwrap edition safely without panicking here ?

P-E-P commented on src/header.rs:
```diff
@@ -0,0 +1,96 @@
+
+    if let Some(after_open_bracket) = after_comment.strip_prefix('[') {
+        let Some((line_revision, directive)) = after_open_bracket.split_once(']') else {
+            panic!(
+                "malformed condition directive: expected `{comment}[foo]`, found `{original_line}`"
+            )
```
I'm not thrilled about this panic here

MahadMuhammad replied:
Sure.
 
~I tried to replicate the compile test tool [`src/tools/compiletest/src/header.rs`](https://github.com/rust-lang/rust/blob/d4822c2d84c242cc7403118b50c571464f38ef8f/src/tools/compiletest/src/header.rs#L727)~

P-E-P commented on src/known-directives.rs:
```diff
@@ -0,0 +1,233 @@
+// Copied from https://github.com/rust-lang/rust/blob/master/src/tools/compiletest/src/command-list.rs
```
We want this source to appear in the documentation

P-E-P commented on src/main.rs:
```diff
@@ -1,8 +1,13 @@
-use anyhow::{Context, Result};
-use clap::Parser;
+//! The main entry point of the program.
```
No need to comment main.rs like this :) File is called `main.rs` and has a main function it is fairly obvious  is the main entry point.

P-E-P commented on src/transform.rs:
```diff
@@ -1,46 +1,66 @@
+            // TODO: This is not the efficient way to find respective line number
+            for error in errors.iter() {
+                // Checking the original line number
+                if (error.line_num as i32 - error.relative_line_num) != line_num as i32 {
```
I'd prefer if you could use either the `From` trait or the `TryFrom` trait (or their `Into` counterpart) instead of a cast because I don't like silent errors that morph into bugs.

MahadMuhammad replied:
Sure, this surely helps to catch undesired behaviors :)

P-E-P reviewed (approved):
A lot of great things in there, most of my comments are nitpicks because you did really great :laughing:

MahadMuhammad committed:
chore/cleanup: refactor code and comments

- in `cli`
	-  converted doc comments to regular comments
- in `errors`
	-  added brief macro/fmt for `Error` documentation
	-  Add proper indentation and markdown annotation
- in `header`
	-  Called `unwrap` safely
	-  remove unwanted panic, initially this was the
	   replication of rust compile test tool, but
	   this, wasn't unwanted here. See compile test
	   internals https://github.com/rust-lang/rust/blob/d4822c2d84c242cc7403118b50c571464f38ef8f/src/tools/compiletest/src/header.rs#L727
- in `known-directives`:
	-  Add directives link in documentation
- in `main`:
	- remove `unwanted` entry point comment
- in `transform`:
	- Add `TryFrom` for line number casting

Signed-off-by: Muhammad Mahad <mahadtxt@gmail.com>

MahadMuhammad committed:
refactor: compiling `regex` outside loop

MahadMuhammad commented:
Thanks for review @P-E-P @CohenArthur, I'll update further PR's once this one will be merged :)
